### PR TITLE
Set to Next primitive operator "object with same semiglobal as"

### PR DESCRIPTION
**Pull request by andrewknoll**
The primitive Set To Next allows for a new operator called "object with same semiglobal as", which iterates through any object whose semiglobal is shared with the object type that has been chosen.

### CODE REVIEW

**Comment by TheArchitectFreeSO** [comment]
This is really great. I can think of several instances that I'd like to use this someday. Thanks!

**Review by riperiperi — APPROVED** [review]
Looks good, but you'll want to be careful with identifiers for new features. 

Since this is an FSO specific feature, you should call it `FSOObjectOfSemiGlobal`, and have its value notably far away from the maxis set to next modes. (like 255 or something, rather than 13 immediately after the TS1 ones)

File: TSOClient/tso.simantics/Model/VMObjectQueries.cs [hunk 1]
```diff
@@ -156,6 +187,14 @@ public void NewObject(VMEntity obj)
             }
             VM.AddToObjList(list, obj);
             RegisterCategory(obj, obj.GetValue(VMStackObjectVariable.Category));
+            if(obj.SemiGlobal != null)
+            {
+                if(obj.SemiGlobal.Iff.Filename != null)   //sanity check
```
Comment (riperiperi) [inline TSOClient/tso.simantics/Model/VMObjectQueries.cs:192]: `if (`
If needs space

Also, what's with all the spaces before `//sanity check`?

File: TSOClient/tso.simantics/Model/VMObjectQueries.cs [hunk 2]
```diff
@@ -180,6 +219,14 @@ public void RemoveObject(VMEntity obj)
                 if (list.Count == 0) ObjectsByGUID.Remove(guid);
             }
             RemoveCategory(obj, obj.GetValue(VMStackObjectVariable.Category));
+            if (obj.SemiGlobal != null)
+            {
+                if (obj.SemiGlobal.Iff.Filename != null)    //sanity check
```
Comment (riperiperi) [inline TSOClient/tso.simantics/Model/VMObjectQueries.cs:224]: Too many spaces before `//sanity check`

File: TSOClient/tso.simantics/Primitives/VMSetToNext.cs [hunk 3]
```diff
@@ -92,6 +92,20 @@ public override VMPrimitiveExitCode Execute(VMStackFrame context, VMPrimitiveOpe
                         entities = context.VM.Context.ObjectQueries.GetObjectsByGUID(operand.GUID); break;
                     case VMSetToNextSearchType.ObjectWithCategoryEqualToSP0:
                         entities = context.VM.Context.ObjectQueries.GetObjectsByCategory(context.Args[0]); break;
+                    case VMSetToNextSearchType.FSOObjectOfSemiGlobal:
+                        var sg = FSO.Content.Content.Get().WorldObjects.Get(operand.GUID).Resource.SemiGlobal;
+
+                        if (sg != null)
+                        {
+                            string sg_name = sg.Iff.Filename;   //retrieve semiglobal's iff filename
```
Comment (riperiperi) [inline TSOClient/tso.simantics/Primitives/VMSetToNext.cs:100]: Too many spaces before comment

**Review by riperiperi — CHANGES_REQUESTED** [review]
Looks good, just a few code style concerns.

File: TSOClient/tso.simantics/Model/VMObjectQueries.cs (hunk 1 above)
Comment (andrewknoll) [reply TSOClient/tso.simantics/Model/VMObjectQueries.cs:192]: Not quite sure why I'd put three spaces there, but will fix it asap.

File: TSOClient/tso.simantics/Primitives/VMSetToNext.cs (hunk 3 above)
Comment (andrewknoll) [reply TSOClient/tso.simantics/Primitives/VMSetToNext.cs]: Solved in fc27d4be1bda305e6d4dc490743be9e8b6c21285

File: TSOClient/tso.simantics/Model/VMObjectQueries.cs (hunk 2 above)
Comment (andrewknoll) [reply TSOClient/tso.simantics/Model/VMObjectQueries.cs]: Solved in fc27d4be1bda305e6d4dc490743be9e8b6c21285

File: TSOClient/tso.simantics/Model/VMObjectQueries.cs (hunk 1 above)
Comment (andrewknoll) [reply TSOClient/tso.simantics/Model/VMObjectQueries.cs]: Solved in fc27d4be1bda305e6d4dc490743be9e8b6c21285

File: TSOClient/tso.simantics/Model/VMObjectQueries.cs [hunk 4]
```diff
@@ -156,6 +187,14 @@ public void NewObject(VMEntity obj)
             }
             VM.AddToObjList(list, obj);
             RegisterCategory(obj, obj.GetValue(VMStackObjectVariable.Category));
+            if(obj.SemiGlobal != null)
```
[the remark this answers is not in the document]
Comment (andrewknoll) [reply TSOClient/tso.simantics/Model/VMObjectQueries.cs]: Solved in fc27d4be1bda305e6d4dc490743be9e8b6c21285

File: TSOClient/tso.simantics/Model/VMObjectQueries.cs [hunk 5]
```diff
@@ -144,6 +146,35 @@ public void RemoveCategory(VMEntity obj, short category)
             if (tile.Count == 0) ObjectsByCategory.Remove(category);
         }
 
+        public void RegisterSemiGlobal(VMEntity obj, string semiGlobal)
+        {
+            List<VMEntity> tile = null;
+            if(semiGlobal != null)
```
[the remark this answers is not in the document]
Comment (andrewknoll) [reply TSOClient/tso.simantics/Model/VMObjectQueries.cs]: Solved in fc27d4be1bda305e6d4dc490743be9e8b6c21285

File: TSOClient/FSO.SimAntics.JIT/Translation/CSharp/Primitives/CSSetToNextPrimitive.cs [hunk 6]
```diff
@@ -96,6 +96,14 @@ public override List<string> CodeGen(TranslationContext context)
                         csClass.UseParams = true;
                         codeResult.Add($"var entities = context.VM.Context.ObjectQueries.GetObjectsByCategory(args[0]);");
                         break;
+                    case VMSetToNextSearchType.FSOObjectOfSemiGlobal:
+                        codeResult.Add($"var semiglobal = FSO.Content.Content.Get().WorldObjects.Get(operand.GUID).Resource.SemiGlobal;");
+                        codeResult.Add($"if(sg != null) {{ ");
+                        codeResult.Add($"string sg_name = sg.Iff.Filename;");
+                        codeResult.Add($"if(obj.SemiGlobal.Iff.Filename != null) entities = context.VM.Context.ObjectQueries.GetObjectsBySemiGlobal(sg_name);");
```
[the remark this answers is not in the document]
Comment (andrewknoll) [reply TSOClient/FSO.SimAntics.JIT/Translation/CSharp/Primitives/CSSetToNextPrimitive.cs]: Solved in fc27d4be1bda305e6d4dc490743be9e8b6c21285

File: TSOClient/FSO.SimAntics.JIT/Translation/CSharp/Primitives/CSSetToNextPrimitive.cs [hunk 7]
```diff
@@ -96,6 +96,14 @@ public override List<string> CodeGen(TranslationContext context)
                         csClass.UseParams = true;
                         codeResult.Add($"var entities = context.VM.Context.ObjectQueries.GetObjectsByCategory(args[0]);");
                         break;
+                    case VMSetToNextSearchType.FSOObjectOfSemiGlobal:
+                        codeResult.Add($"var semiglobal = FSO.Content.Content.Get().WorldObjects.Get(operand.GUID).Resource.SemiGlobal;");
+                        codeResult.Add($"if(sg != null) {{ ");
```
[the remark this answers is not in the document]
Comment (andrewknoll) [reply TSOClient/FSO.SimAntics.JIT/Translation/CSharp/Primitives/CSSetToNextPrimitive.cs]: Solved in fc27d4be1bda305e6d4dc490743be9e8b6c21285

**Review by riperiperi — APPROVED** [review]

**Pull request merged**